### PR TITLE
Refactors, performance improvements, and fixes

### PR DESCRIPTION
**Pull request by pauloue**
Benchmark in dreambuilder:

master:
```
[craftguide] Getting init_items
[craftguide] Found 1722 init_items (16592 reg_items) in 16.0126 sec
[craftguide] Running benchmark
get_item_usages
  11.6449
  10.9920
  13.0139
  16.5721
  12.1848
  11.5292
  12.1791
  16.7743
  10.9867
  11.3969
  10.8906
  12.3318
  16.0794
  11.1595
  11.8558
  avg: 12.6394 ms
get_progressive_items
  21.6717
  18.7699
  22.8427
  23.1046
  29.9226
  17.1253
  17.7711
  26.0294
  24.1830
  32.0850
  17.3702
  17.9929
  26.0571
  22.5991
  30.8307
  avg: 23.2237 ms
```
PR:
```
[craftguide] Getting init_items
[craftguide] Found 1722 init_items (16592 reg_items) in 16.2989 sec
[craftguide] Running benchmark
get_item_usages
  0.9279
  0.8491
  0.8352
  1.0106
  1.1313
  1.1292
  1.1588
  1.0636
  0.8733
  1.1002
  1.0572
  1.0998
  1.1199
  1.1237
  0.8954
  avg: 1.0250 ms
get_progressive_items
  7.2115
  12.0551
  8.6089
  7.4014
  8.0327
  19.7317
  8.1721
  7.8774
  28.4355
  19.8928
  6.6494
  6.1713
  6.2526
  6.8371
  6.5223
  avg: 10.6568 ms
```
There's no significant difference in the performance of `get_init_items`, `get_item_usages("default:stick")` is ~92% faster, and `get_progressive_items` is ~54% faster.

To test, the benchmark commit can be cherry-picked from my master branch.
Fuels are cached because `get_craft_result` is very slow (try commenting out `cache_fuel` in `get_init_items`).

Also fixes shapeless icon not showing and the wrong tooltip being shown before that.
Sorry it's so big.

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -986,3 +933,36 @@ for x = 1, 6 do
 	})
 end
 ]]
+
+minetest.register_on_mods_loaded(function()
```
[the remark this answers is not in the document]
Comment (pauloue) [reply init.lua]: `register_on_mods_loaded` is already used elsewhere and the currently released version is 5.0-only. Also this part will probably be dropped as it's just for testing.

File: init.lua [hunk 2]
```diff
@@ -78,46 +80,32 @@ craftguide.register_craft({
...
-			return custom_craft
-		end
-	end
-
-	return {}
-end
-
-local function get_recipes(output)
+local function cache_recipes(output)
 	local recipes = mt.get_all_craft_recipes(output) or {}
```
[the remark this answers is not in the document]
Comment (pauloue) [reply init.lua]: That's the function that adds an item's recipes to the cache and it's only called once per item in `get_init_items`.

File: init.lua [hunk 3]
```diff
@@ -78,46 +80,32 @@ craftguide.register_craft({
...
-local color_codes = {
-	red = "#FF0000",
-	yellow = "#FFFF00",
-}
-
-local function colorize(str, color)
-	color = color or "yellow"
-	return mt.colorize(color_codes[color], str)
+local function get_burntime(item)
+	return get_result({method="fuel", width=1, items={item}}).time
```
Comment (kilbith) [inline init.lua]: Spaces around the operators.

File: init.lua [hunk 4]
```diff
@@ -154,85 +142,60 @@ local function groups_to_item(groups)
...
+		local groupstr = {}
 		for i = 1, #groups do
-			groupstr = groupstr ..
-				colorize(groups[i]) .. (groups[i + 1] and ", " or "")
+			groupstr[#groupstr + 1] = mt.colorize("yellow", groups[i])
 		end
-
-		tooltip = tooltip ..
-			S("Any item belonging to the group(s)") .. ": " .. groupstr
+		groupstr = concat(groupstr, ",")
```
Comment (kilbith) [inline init.lua]: `, `

File: init.lua [hunk 5]
```diff
@@ -266,35 +228,40 @@ local function get_recipe_fs(iX, iY, xoffset, recipe_num, recipes, show_usage)
...
 			rightest + 1.2,
 			iY + (sfinv_only and 2.2 or 1.7),
 			0.5,
 			0.5,
-			custom_recipe and custom_recipe.icon or
-				"craftguide_" .. icon .. ".png^[resize:16x16")
+			icon)
+
+		local tooltip = custom_recipe and custom_recipe.description or
+				shapeless and "Shapeless" or "Cooking"
```
Comment (kilbith) [inline init.lua]: `S("Shapeless") or S("Cooking")`

File: init.lua [hunk 6]
```diff
@@ -305,46 +272,59 @@ local function get_recipe_fs(iX, iY, xoffset, recipe_num, recipes, show_usage)
...
+		data.show_usages and S("Usage") or S("Recipe"),
+		data.rnum,
+		S("of"),
+		#data.recipes)
 
 	return concat(fs)
 end
 
-local function make_formspec(player_name)
+local function get_formspec(player_name)
```
Comment (kilbith) [inline init.lua]: Keep `make_formspec`

File: init.lua [hunk 7]
```diff
@@ -377,15 +357,13 @@ local function make_formspec(player_name)
 	fs[#fs + 1] = "image_button[" .. (data.iX - (sfinv_only and 2.6 or 3.1)) ..
 			",0.12;0.8,0.8;craftguide_prev_icon.png;prev;]"
 	fs[#fs + 1] = "label[" .. (data.iX - (sfinv_only and 1.7 or 2.2)) ..
-			",0.22;" .. colorize(data.pagenum) .. " / " .. data.pagemax .. "]"
+			",0.22;" .. mt.colorize("yellow", data.pagenum) .. " / " .. data.pagemax .. "]"
```
Comment (kilbith) [inline init.lua]: Line too long (> 100 col.)

File: init.lua [hunk 8]
```diff
@@ -489,63 +454,63 @@ end
...
+		for i = 1, #recipes do
+			local recipe = recipes[i]
+			if item_in_recipe(item, recipe) then
+				usages[#usages + 1] = recipe
 			end
 		end
 	end
 
+	if fuel_cache[item] then
+		usages[#usages + 1] = {type="fuel", width=1, items={item}}
```
Comment (kilbith) [inline init.lua]: Spaces around operators.

File: init.lua [hunk 9]
```diff
@@ -489,63 +454,63 @@ end
...
+	end
+
 	return usages
 end
 
 local function get_inv_items(player)
-	local invlist = player:get_inventory():get_list("main")
+	local lists = player:get_inventory():get_lists()
```
Comment (kilbith) [inline init.lua]: Is it faster than getting the `main` list directly or there are other reasons to do this?

File: init.lua [hunk 10]
```diff
@@ -489,63 +454,63 @@ end
...
+				return true
 			end
-		else
-			for i = 1, inv_items_size do
-				if inv_items[i] == item then
-					item_in_inv = true
-				end
+		end
+	else
+		for i = 1, #inv_items do
```
Comment (kilbith) [inline init.lua]: Localize `#inv_items`

File: init.lua [hunk 11]
```diff
@@ -620,53 +587,47 @@ end
...
+	data.pagenum     = 1
+	data.query_item  = nil
+	data.show_usages = nil
+	data.recipes     = nil
+	data.items       = progressive_mode and data.progressive_items or init_items
 end
 
 local function get_init_items()
+	print("[craftguide] Getting init_items")
+	local start, ct = os.clock(), 0
```
Comment (kilbith) [inline init.lua]: Please remove all the benchmark code, I don't need this.

File: init.lua [hunk 12]
```diff
@@ -714,66 +673,62 @@ local function get_fields(player, ...)
 	elseif (fields.size_inc and data.iX < MAX_LIMIT) or
 			(fields.size_dec and data.iX > MIN_LIMIT) then
 		data.pagenum = 1
-		data.iX = data.iX - (fields.size_dec and 1 or -1)
+		data.iX = data.iX + (fields.size_inc and 1 or -1)
```
Comment (kilbith) [inline init.lua]: Was that necessary?...

File: init.lua [hunk 13]
```diff
@@ -714,66 +673,62 @@ local function get_fields(player, ...)
...
-				item = item:sub(1,-5)
+	else
+		local item
+		for field in pairs(fields) do
+			if field:find(":") then
+				item = field
+				break
 			end
+		end
+		if not item then return end
```
Comment (kilbith) [inline init.lua]:
```Lua
... then
        return
end
```

File: init.lua (hunk 12 above)
Comment (pauloue) [reply init.lua]: No. It just felt backward

File: init.lua (hunk 6 above)
Comment (pauloue) [reply init.lua]: I thought `get_formspec` would make more sense now that it always returns the formspec. Also there's already `sfinv.make_formspec` and it's consistent with `get_recipe_fs`.

File: init.lua (hunk 10 above)
Comment (pauloue) [reply init.lua]: It's in a separate function now so would require an additional parameter. Also I really think that would not improve performance in any measurable way.

File: init.lua (hunk 9 above)
Comment (pauloue) [reply init.lua]: It's so items in the craft grid are included. Thought it should be just `main` and `craft`, not the others.

File: init.lua (hunk 10 above)
Comment (kilbith) [reply init.lua]: Didn't see it's in a separate function, nevermind then.

File: init.lua (hunk 6 above)
Comment (kilbith) [reply init.lua]: Keep `make_formspec` anyway. `get` feels like "showing", we are only "building" a formspec here.

File: init.lua [hunk 14]
```diff
@@ -489,63 +454,63 @@ end
...
+	end
+
 	return usages
 end
 
 local function get_inv_items(player)
-	local invlist = player:get_inventory():get_list("main")
+	local inv = player:get_inventory()
+	local stacks = inv:get_list("main")
+	table.insert_all(stacks, inv:get_list("craft"))
```
Comment (kilbith) [inline init.lua]: Don't rely on a core helper for such a simple function. Do it locally.

File: init.lua [hunk 15]
```diff
@@ -620,31 +587,31 @@ end
 local function init_data(player, name)
 	local p_items = progressive_mode and get_progressive_items(player) or nil
 	player_data[name] = {
+		items   = p_items or init_items,
+		progressive_items = p_items,
```
Comment (kilbith) [inline init.lua]: Keep that at the bottom of the table.

**Comment by kilbith** [comment]
Merged, thanks.

https://github.com/minetest-mods/craftguide/commit/d950c71dbc8ba05abef44f8a553e65e2a7b4157f

**Pull request closed without merging**